src/tagifyme/model/Database.java

```diff
@@ -197,24 +198,56 @@ public Iterable<Pair<Data, Set<Tag>>> data_and_tags() {
     return r;
   }
   
-  public Iterable<Tag> tags(){
+  /**
+   * Return the `Tag`s within the `Database`.
+   */
+  public Iterable<Tag> allTags(){
       return tag_set;
   }
-
   
+  // TODO: I don't know why this is needed.
+  // This might just be a `contains`?
   public Tag getTag(String name){
     int i = 0;
     for (Tag tag : tag_set ){
         if(tag.getName().equals(name)){
             return tag;
         };
     }
-    return noTag;
+    return new Tag("Undefined");
   }
   
+  /**
+   * Delete some Data from the Database by PATH.
+   */
+  public void deleteDataByPATH(String PATH) {
+    // TODO: I have no idea why you've chosen to delete
+    // via PATH; it's likely easier to do it with `name`
+    // only because there will be 'Set' contains type
+    // methods; but whatever.
+    for (Data d : data_set) {
+      if (d.getPATH().equals(PATH)) {
+        // TODO: Again, is there any risk to modifying the
+        // underlying structure while iterating over it?
```

Comment:
idk possibly, we can just move it out of the loop

Reply:
I don't really think it's an issue: within some languages modifying the underlying structure of the thing being iterated over might lead to undefined behavior. See [this](https://langdev.stackexchange.com/questions/2593/what-approaches-are-there-to-prevent-modifying-a-collection-while-iterating-over). I don't think it matters in something that's managed/has a GC/is developed as Java.

At the same time, if it breaks, it breaks.